File: worker/src/formatter.js
// JS-форматтер сводки. Зеркалит bot/formatter.py байт-в-байт.

const NBSP = " "; // обычный пробел (исторический формат бота)
const SEP = "───────────────";

export function fmtInt(n) {
  const x = Math.trunc(n);
  const abs = Math.abs(x);
  const s = String(abs).replace(/\B(?=(\d{3})+(?!\d))/g, NBSP);
  return x < 0 ? "−" + s : s;
}

export function fmtPct(x) {
  return (x * 100).toFixed(1).replace(".", ",") + "%";
}

function planTail(p, withRub = false) {
  if (!p || p.plan == null) return "";
  const rub = withRub ? `${NBSP}₽` : "";
  return ` из ${fmtInt(p.plan)}${rub} (${fmtPct(p.pct_of_plan)}) · RR ${fmtPct(p.run_rate)}`;
}

function formatDaily(day, header) {
  const out = [header, ""];
  const dly = day.daily;
  out.push(`• Новых клиентов: ${fmtInt(dly.new_clients)}`);
  out.push(`• Новых карт: ${fmtInt(dly.new_cards.total)}`);
  out.push(`   – Travel: ${fmtInt(dly.new_cards.travel)}`);
  out.push(`   – Premium: ${fmtInt(dly.new_cards.premium)}`);
  out.push(`   – Subscriptions: ${fmtInt(dly.new_cards.subscriptions)}`);
  out.push("");
  out.push(`• GMV: ${fmtInt(dly.gmv.total)}${NBSP}₽`);
  out.push(`   – за выпуск карт: ${fmtInt(dly.gmv.cards_issue)}${NBSP}₽`);
  out.push(`   – за пополнения: ${fmtInt(dly.gmv.topups)}${NBSP}₽`);
  out.push("");
  out.push(`• Revenue: ${fmtInt(dly.revenue.total)}${NBSP}₽`);
  out.push(`   – с открытия карт: ${fmtInt(dly.revenue.cards_issue)}${NBSP}₽`);
  out.push(`   – с пополнений: ${fmtInt(dly.revenue.topups)}${NBSP}₽`);
  out.push("");
  const gp = dly.gross_profit;
  out.push(`• Gross Profit: ${fmtInt(gp.total)}${NBSP}₽`);
  if (gp.cards_issue != null) out.push(`   – GP с открытия карт: ${fmtInt(gp.cards_issue)}${NBSP}₽`);
  if (gp.topups != null) out.push(`   – GP с пополнений: ${fmtInt(gp.topups)}${NBSP}₽`);
  out.push(`   – OpEx: ${fmtInt(gp.opex)}${NBSP}₽`);
  out.push(`   – ComEx: ${fmtInt(gp.comex)}${NBSP}₽`);
  out.push("");
  out.push(`• Чистая прибыль: ${fmtInt(dly.net_profit)}${NBSP}₽`);
  return out;
}

function formatPeriod(p) {
  const out = ["", SEP, "", `📊 За ${p.label}`, ""];
  const nc = p.new_clients;
  out.push(`• Новых клиентов: ${fmtInt(nc.actual)}${planTail(nc)}`);
  const cards = p.new_cards;
  out.push(`• Новых карт: ${fmtInt(cards.actual)}${planTail(cards)}`);
  out.push(`   – Travel: ${fmtInt(cards.travel)}`);
  out.push(`   – Premium: ${fmtInt(cards.premium)}`);
  out.push(`   – Subscriptions: ${fmtInt(cards.subscriptions)}`);
  out.push("");
  const gmv = p.gmv;
  out.push(`• GMV: ${fmtInt(gmv.actual)}${NBSP}₽${planTail(gmv, true)}`);
  out.push(`   – за выпуск карт: ${fmtInt(gmv.cards_issue)}${NBSP}₽`);
  out.push(`   – за пополнения: ${fmtInt(gmv.topups)}${NBSP}₽`);
  out.push("");
  const rev = p.revenue;
  out.push(`• Revenue: ${fmtInt(rev.actual)}${NBSP}₽`);
  out.push(`   – с открытия карт: ${fmtInt(rev.cards_issue)}${NBSP}₽`);
  out.push(`   – с пополнений: ${fmtInt(rev.topups)}${NBSP}₽`);
  out.push("");
  const gp = p.gross_profit;
  out.push(`• Gross Profit: ${fmtInt(gp.actual)}${NBSP}₽${planTail(gp, true)}`);
  if (gp.cards_issue != null) out.push(`   – GP с открытия карт: ${fmtInt(gp.cards_issue)}${NBSP}₽`);
  if (gp.topups != null) out.push(`   – GP с пополнений: ${fmtInt(gp.topups)}${NBSP}₽`);
  const daysTail = p.days_elapsed ? ` (за ${p.days_elapsed}${NBSP}дн.)` : "";
  out.push(`   – OpEx: ${fmtInt(gp.opex)}${NBSP}₽${daysTail}`);
  out.push(`   – ComEx: ${fmtInt(gp.comex)}${NBSP}₽${daysTail}`);
  out.push("");
  const np = p.net_profit;
  out.push(`• Чистая прибыль: ${fmtInt(np.actual)}${NBSP}₽${planTail(np, true)}`);
  return out;
}

export function formatSummary(day) {
  const [y, m, d] = day.date.split("-");
  const header = `Ежедневная сводка ППМ — ${d}.${m}.${y}`;
  let lines = formatDaily(day, header);
  if (day.month) lines = lines.concat(formatPeriod(day.month));
  if (day.quarter) lines = lines.concat(formatPeriod(day.quarter));
  return lines.join("\n");
}

File: bot/formatter.py
"""
Форматирование сводки в текст Telegram-сообщения.

Контракт совпадает 1:1 с историческими сообщениями бота (data/raw_messages.txt).
"""
from __future__ import annotations

from datetime import date

NBSP = " "
SEP = "───────────────"


def fmt_int(n: int) -> str:
    s = f"{abs(int(n)):,}".replace(",", NBSP)
    return ("−" + s) if n < 0 else s


def fmt_pct(x: float) -> str:
    return f"{x * 100:.1f}".replace(".", ",") + "%"


def _plan_tail(p: dict, with_rub: bool = False) -> str:
    if not p or p.get("plan") is None:
        return ""
    rub = f"{NBSP}₽" if with_rub else ""
    return (
        f" из {fmt_int(p['plan'])}{rub}"
        f" ({fmt_pct(p['pct_of_plan'])})"
        f" · RR {fmt_pct(p['run_rate'])}"
    )


def _format_daily(d: dict, header: str) -> list[str]:
    out = [header, ""]
    dly = d["daily"]
    out.append(f"• Новых клиентов: {fmt_int(dly['new_clients'])}")
    out.append(f"• Новых карт: {fmt_int(dly['new_cards']['total'])}")
    out.append(f"   – Travel: {fmt_int(dly['new_cards']['travel'])}")
    out.append(f"   – Premium: {fmt_int(dly['new_cards']['premium'])}")
    out.append(f"   – Subscriptions: {fmt_int(dly['new_cards']['subscriptions'])}")
    out.append("")
    out.append(f"• GMV: {fmt_int(dly['gmv']['total'])}{NBSP}₽")
    out.append(f"   – за выпуск карт: {fmt_int(dly['gmv']['cards_issue'])}{NBSP}₽")
    out.append(f"   – за пополнения: {fmt_int(dly['gmv']['topups'])}{NBSP}₽")
    out.append("")
    out.append(f"• Revenue: {fmt_int(dly['revenue']['total'])}{NBSP}₽")
    out.append(f"   – с открытия карт: {fmt_int(dly['revenue']['cards_issue'])}{NBSP}₽")
    out.append(f"   – с пополнений: {fmt_int(dly['revenue']['topups'])}{NBSP}₽")
    out.append("")
    gp = dly["gross_profit"]
    out.append(f"• Gross Profit: {fmt_int(gp['total'])}{NBSP}₽")
    if "cards_issue" in gp:
        out.append(f"   – GP с открытия карт: {fmt_int(gp['cards_issue'])}{NBSP}₽")
    if "topups" in gp:
        out.append(f"   – GP с пополнений: {fmt_int(gp['topups'])}{NBSP}₽")
    out.append(f"   – OpEx: {fmt_int(gp['opex'])}{NBSP}₽")
    out.append(f"   – ComEx: {fmt_int(gp['comex'])}{NBSP}₽")
    out.append("")
    out.append(f"• Чистая прибыль: {fmt_int(dly['net_profit'])}{NBSP}₽")
    return out


def _format_period(p: dict) -> list[str]:
    out = ["", SEP, "", f"📊 За {p['label']}", ""]
    nc = p["new_clients"]
    out.append(f"• Новых клиентов: {fmt_int(nc['actual'])}{_plan_tail(nc)}")
    cards = p["new_cards"]
    out.append(f"• Новых карт: {fmt_int(cards['actual'])}{_plan_tail(cards)}")
    out.append(f"   – Travel: {fmt_int(cards['travel'])}")
    out.append(f"   – Premium: {fmt_int(cards['premium'])}")
    out.append(f"   – Subscriptions: {fmt_int(cards['subscriptions'])}")
    out.append("")
    gmv = p["gmv"]
    out.append(f"• GMV: {fmt_int(gmv['actual'])}{NBSP}₽{_plan_tail(gmv, with_rub=True)}")
    out.append(f"   – за выпуск карт: {fmt_int(gmv['cards_issue'])}{NBSP}₽")
    out.append(f"   – за пополнения: {fmt_int(gmv['topups'])}{NBSP}₽")
    out.append("")
    rev = p["revenue"]
    out.append(f"• Revenue: {fmt_int(rev['actual'])}{NBSP}₽")
    out.append(f"   – с открытия карт: {fmt_int(rev['cards_issue'])}{NBSP}₽")
    out.append(f"   – с пополнений: {fmt_int(rev['topups'])}{NBSP}₽")
    out.append("")
    gp = p["gross_profit"]
    out.append(f"• Gross Profit: {fmt_int(gp['actual'])}{NBSP}₽{_plan_tail(gp, with_rub=True)}")
    if "cards_issue" in gp:
        out.append(f"   – GP с открытия карт: {fmt_int(gp['cards_issue'])}{NBSP}₽")
    if "topups" in gp:
        out.append(f"   – GP с пополнений: {fmt_int(gp['topups'])}{NBSP}₽")
    days = p.get("days_elapsed")
    days_tail = f" (за {days}{NBSP}дн.)" if days else ""
    out.append(f"   – OpEx: {fmt_int(gp['opex'])}{NBSP}₽{days_tail}")
    out.append(f"   – ComEx: {fmt_int(gp['comex'])}{NBSP}₽{days_tail}")
    out.append("")
    np_ = p["net_profit"]
    out.append(f"• Чистая прибыль: {fmt_int(np_['actual'])}{NBSP}₽{_plan_tail(np_, with_rub=True)}")
    return out


def format_summary(day: dict) -> str:
    d = date.fromisoformat(day["date"])
    header = f"Ежедневная сводка ППМ — {d.strftime('%d.%m.%Y')}"
    lines = _format_daily(day, header)
    if day.get("month"):
        lines += _format_period(day["month"])
    if day.get("quarter"):
        lines += _format_period(day["quarter"])
    return "\n".join(lines)
